### ISSUE TITLE
Logging to Cloudwatch from PHP-FPM

### ISSUE THREAD

**Issue by mnapoli** [mnapoli/bref#184]
We should find a solution (and document it) to log to Cloudwatch from PHP-FPM processes.

Ideally it should be without going through a file (e.g. in `/tmp`). Maybe with `error_log`? Gotta try this out.

**Comment by sandrokeil** [comment]
Maybe this works like in Docker.

`php-fpm.conf`
```
[global]

error_log = /proc/self/fd/2
```

`pool.conf`
```
[www]

; Ensure worker stdout and stderr are sent to the main error log.
catch_workers_output = yes
```

**Comment by mnapoli** [comment]
Lambda isn't running Docker though, they are running their own implementation called Firecracker.

**Comment by mnapoli** [comment]
So far I think I have found 2 solutions.

First in both cases we need to redirect PHP-FPM's output to stderr:

```diff
-        $this->fpm = new Process(['php-fpm', '--nodaemonize', '--fpm-config', $this->configFile]);
+        $this->fpm = new Process(['php-fpm', '--nodaemonize', '--force-stderr', '--fpm-config', $this->configFile]);
        $this->fpm->setTimeout(null);
        $this->fpm->start(function ($type, $output): void {
-            if ($type === Process::ERR) {
-                echo $output;
-                exit(1);
-            }
+            echo $output;
        });
```

### Log in `stderr` through PHP-FPM

In `php-fpm.conf`:

```ini
catch_workers_output = yes
```

Then in PHP scripts we can write to `stderr`:

```php
$f = fopen('php://stderr', 'a+');
fwrite($f, 'My log');
fclose($f);
```

However this shows up like this in Cloudwatch:

```
[19-Jan-2019 15:41:24] NOTICE: fpm is running, pid 14
[19-Jan-2019 15:41:24] NOTICE: ready to handle connections
[19-Jan-2019 15:41:25] WARNING: [pool default] child 15 said into stderr: "My log"
```

That works but the log line is pretty long. Imagine a Monolog line with the default config, it would look like this in cloudwatch:

```
2019-01-19T13:52:57.167Z	c423d3a3-e975-484f-bbcb-f0d0c6865ee9	[STDERR] [19-Jan-2019 15:41:25] WARNING: [pool default] child 15 said into stderr: [2018-07-05 11:18:29] default.INFO: My log
```

That could probably be simplified however.

## Log using `error_log`

In PHP:

```php

```

However this echos the log message in the FPM response on the FastCGI socket. This is the full HTTP response sent by the PHP-FPM worker:

```
PHP message: My log
X-Powered-By: PHP/7.2.14
Content-type: text/html; charset=UTF-8

Hello world!
```

You'll notice the first line **is not a HTTP header**, it's the output of `error_log` that was inserted in the HTTP response.

That means that HOA FastCGI parses the HTTP response incorrectly. We could however change how the HTTP response is parsed, but I'm not sure we could do that reliably.

I'm not convinced about this solution.

## Was doesn't work

I wasn't able to tell if `/dev/stderr` can be used in Lambda, for now I'm running my tests locally and this file is usable but this is not the same thing as on AWS Lambda (locally SAM uses Docker).

The next step is probably try to simply use `/dev/stderr`, deploy to lambda and see if it works. That would be a much simpler solution than trying to forward everything from the PHP-FPM worker to the parent PHP-FPM process to the parent PHP `bootstrap` script…

**Comment by mnapoli** [comment]
Actually `/dev/stderr` doesn't work at all with `sam local` so I won't even try it in Lambda.

`error_log()` doesn't seem like a bad solution but:

- it means adjusting the Hoa FastCGI client library to parse both the HTTP response and the preceding error log messages (I wonder how it works with Apache/Nginx…)
- it seems that the HTTP response is parsed by Hoa at once, that means the logs will not be transmitted as they arrive to AWS -> if the Lambda function times out then I'm afraid we'll loose some logs

**Comment by mnapoli** [comment]
Pinging @bukka, maybe you can give us some advices here on how to log from PHP-FPM workers to the parent process of the PHP-FPM master.

Here is a summary of how it works with PHP on Lambda:

```
PHP bootstrap file
  -> runs `php-fpm --nodaemonize --force-stderr`
    -> runs the php-fpm workers
       the PHP script handles a HTTP request and tries to log on stderr
```

The log from the PHP script (PHP-FPM worker) must go back to the PHP bootstrap file. The question is *how can we do that?*

**Comment by mnapoli** [comment]
I've just realized that in https://github.com/php/php-src/pull/2458 the feature to disable the prefix in the PHP-FPM log has actually been implemented 🎉 

This is the new `decorate_workers_output` option released [in PHP 7.3](http://php.net/manual/en/migration73.new-features.php).

So to sum up what we can do is use the `stderr` solution and:

- have an ugly prefix in PHP 7.2
- have a nice log line in PHP 7.3

However we don't support 7.3 yet in Bref ^^ I know it will complicate the build scripts to add it… And we don't know if all the extensions we are installing ATM are compatible yet…
